aiokafka/record/legacy_records.py

```diff
@@ -278,155 +285,154 @@ def __repr__(self):
 class _LegacyRecordBatchBuilderPy(LegacyRecordBase):
 
     def __init__(self, magic, compression_type, batch_size):
+        assert magic in [0, 1]
         self._magic = magic
         self._compression_type = compression_type
         self._batch_size = batch_size
-        self._buffer = bytearray()
+        self._buffer = bytearray(batch_size)
```

Comment:
Not sure about this overalocation... We still do bounding checks anyway, does it give that much speed boost? The problem here is that it's a user specified field... It may be pretty big. This can chunk alot of memory away(

Reply:
We could try to maintain a list of bytearrays instead and join on build. May be faster

Reply:
The performance difference is significant but not huge. There are two costs in extending the bytearray: creating the second bytearray and performing the actual extension. These costs grow with the size of each bytearray. Here's an illustration of the difference:

```python
>>> from timeit import timeit
>>> def oneshot():
...   return bytearray(16384)
...
>>> def multiple():
...  b = bytearray()
...  for _ in range (100):
...   b.extend(bytearray(150))
...
>>> timeit("oneshot()", setup="from __main__ import oneshot, multiple")
0.5214737450005487
>>> timeit("multiple()", setup="from __main__ import oneshot, multiple")
33.52806271798909
```

I don't think memory boundaries are a huge concern. First, each buffer can already take up the specified amount of memory. Second, unless the user is going through the batch interface (in which case they're explicitly taking control anyway) only two batches are allowed per TP. It would take fairly exceptional circumstances to cause noticeable memory pressure for the vast majority of running environments. Even at Netflix's scale (up to 10k partitions per cluster), the default batch settings would max out at ~320MB of memory.

I suspect most use-cases max out at less than 500 partitions, which by default would put this around 16MB of memory. If someone is tweaking their batch size, hopefully they realize that changes their memory use profile. Even if not, though, it's unlikely this will create problems.

Reply:
If I set `max_batch_size=1024 * 1024 * 1024 * 1024` it will get bad. And the name `max_batch_size` does not imply that behaviour. This also differs from the way C extension handles the case, which is not what I want.

Reply:
Is that a common practice?

The C extensions are brand new; I certainly understand wanting to make them behave the same way as the Python code and would be happy to add that diff.